File: src/031/p031.py
coins = [1, 2, 5, 10, 20, 50, 100, 200]


def count_ways(total: int, index: int) -> int:
    if index <= 0:
        return 1

    limit = total // coins[index] + 1
    return sum(count_ways(total - i * coins[index], index - 1) for i in range(limit))


def solve() -> int:
    return count_ways(200, len(coins) - 1)

File: src/031/p031.go
package problem031

import "strconv"

func Solve() string {
	coins := []int{1, 2, 5, 10, 20, 50, 100, 200}

	amount := 200
	ways := make([]int, amount+1)
	ways[0] = 1
	for _, coin := range coins {
		for target := coin; target <= amount; target++ {
			ways[target] += ways[target-coin]
		}
	}
	return strconv.Itoa(ways[amount])
}
